domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,105 @@
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Private mutable class that stores a grade as a fraction.
+ */
+private class MutableFractionGrade(var numerator: Int, var denominator: Int)
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class can exist across multiple training session instances, but calling code is responsible
+ * for ensuring it is properly reset.
+ */
+@Singleton
+internal
+class QuestionAssessmentCalculation @Inject constructor(
+  @ViewHintScorePenalty private val viewHintPenalty: Int,
+  @WrongAnswerScorePenalty private val wrongAnswerPenalty: Int,
+  @MaxScorePerQuestion private val maxScorePerQuestion: Int
+) {
+  private lateinit var skillIdList: List<String>
+  private lateinit var questionSessionMetrics: List<QuestionSessionMetrics>
+  private lateinit var totalScore: MutableFractionGrade
+  private lateinit var scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
+
+  /** Initialize member fields. */
+  internal fun initialize(
+    skillIdList: List<String>,
+    questionSessionMetrics: List<QuestionSessionMetrics>
+  ) {
+    this.skillIdList = skillIdList
+    this.questionSessionMetrics = questionSessionMetrics
+    this.totalScore =
+      MutableFractionGrade(0, 0)
+    createScorePerSkillMapping()
+  }
+
+  /** Initialize the scorePerSkillMapping member field. */
+  private fun createScorePerSkillMapping() {
+    scorePerSkillMapping = mutableMapOf()
+    for (skillId in skillIdList) {
+      scorePerSkillMapping[skillId] = MutableFractionGrade(0, 0)
+    }
+  }
+
+  /** Calculate the user's overall score and score per skill for this practice session. */
+  private fun calculateScores() {
+    for (questionMetric in questionSessionMetrics) {
+      val totalHintsPenalty =
+        questionMetric.numberOfHintsUsed * viewHintPenalty
+      val totalWrongAnswerPenalty =
+        (questionMetric.numberOfAnswersSubmitted - 1) * wrongAnswerPenalty
+      var questionScore = 0
+      if (!questionMetric.didViewSolution) {
+        questionScore =
+          maxOf(
+            maxScorePerQuestion - totalHintsPenalty - totalWrongAnswerPenalty,
+            questionScore
+          )
+      }
+      this.totalScore.numerator += questionScore
+      this.totalScore.denominator += maxScorePerQuestion
+      for (linkedSkillId in questionMetric.question.linkedSkillIdsList) {
+        if (!scorePerSkillMapping.containsKey(linkedSkillId)) continue
+        scorePerSkillMapping[linkedSkillId]!!.numerator += questionScore
+        scorePerSkillMapping[linkedSkillId]!!.denominator += maxScorePerQuestion
+      }
+    }
+  }
+
+  /** Compute the overall score as well as the score and mastery per skill. */
+  internal fun computeAll(): UserAssessmentPerformance {
+    calculateScores()
+    // TODO(#3067): Create mastery calculations method
+
+    // Set up the return values
+    var finalScore = FractionGrade.newBuilder().apply {
```

Comment:
Here & elsewhere: prefer to use val as much as you can. We should almost never need vars. They're harder to make correct since their references can change at any time. For cases where we do need to use mutable classes (like a proto builder or a mutableList/Map), prefer to isolate it just to a single scope & store immutable copies that are shared across a class or across threads.

Reply:
Fixed